### ISSUE TITLE
Using multiple devices in Pytorch Lightning results in multiple ModelCheckpoint callbacks calls

### ISSUE THREAD

**Issue by mrembalski** [Lightning-AI/lightning#17242]
### Outline & Motivation

Pytorch Lightning makes it very easy to use multiple devices. Also makes it very easy to save checkpoints. 

However, both of these things combined, make some undefined behaviour possible. 

### Pitch

1. I think, a warning that a multi-device strategy is used while a model checkpoint callback is also used should be shown.  
2. Docs ([Trainer](https://lightning.ai/docs/pytorch/stable/common/trainer.html)) could be updated to suggest, that for a distributed environment, users should use Trainer.save_checkpoint (just add an URL to [checkpointing](https://lightning.ai/docs/pytorch/stable/common/checkpointing_intermediate.html) page). 
3. Rephrase [Manual saving with distributed training](https://lightning.ai/docs/pytorch/stable/common/checkpointing_intermediate.html):
```
In distributed training cases where a model is running across many machines, Lightning ensures that only one checkpoint is saved instead of a model per machine. This requires no code changes as seen below:
```
Well, it does NOT ensure that in the most common case (Trainer.enable_checkpointing = True). 
4. [possible feature] Maybe make ModelCheckpoint dependent on Strategy and number of devices?

### Additional context

By adding: 
```py
        pid = multiprocessing.current_process().pid
        print(f"{pid}: Checkpointing {filepath}")
```
Inside `_save_checkpoint` function @ ModelCheckpoint, it is possible to see that each of the processes calls the ModelCheckpoint callback.

```py
import os
from torch import optim, nn, utils
from torchvision.datasets import MNIST
from torchvision.transforms import ToTensor
import pytorch_lightning as pl

# define any number of nn.Modules (or use your current ones)
encoder = nn.Sequential(nn.Linear(28 * 28, 64), nn.ReLU(), nn.Linear(64, 3))
decoder = nn.Sequential(nn.Linear(3, 64), nn.ReLU(), nn.Linear(64, 28 * 28))


# define the LightningModule
class LitAutoEncoder(pl.LightningModule):

    def __init__(self, encoder, decoder):
        super().__init__()
        self.encoder = encoder
        self.decoder = decoder

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop.
        # it is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = nn.functional.mse_loss(x_hat, x)
        # Logging to TensorBoard (if installed) by default
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


# init the autoencoder
autoencoder = LitAutoEncoder(encoder, decoder)

# setup data
dataset = MNIST(os.getcwd(), download=True, transform=ToTensor())
train_loader = utils.data.DataLoader(dataset)

# train the model (hint: here are some helpful Trainer arguments for rapid idea iteration)
trainer = pl.Trainer(
    limit_train_batches=100, max_epochs=1, strategy="ddp", devices=4, accelerator="cpu", enable_checkpointing=True)

trainer.fit(model=autoencoder, train_dataloaders=train_loader)
```

cc @justusschock @awaelchli

**Comment by carmocca** [comment]
Copying over the Slack thread:

**william falcon**
last i checked, lightning saves one checkpoint on the root node (that’s been true for years). it does not save checkpoints on the other nodes.  (edited) 

**Michał Rembalski**
Hi 
 - first of all, it’s great to use PL and out of the tools available, I really like it.
The result may indeed be just one file on one node. I just checked that the function _save_checkpoint (which calls trainer.save_checkpoint inside) is actually called multiple times from different processes.
I stumbled upon it by accident (wanted to ensure that metadata of any saved checkpoint would also be saved in another place - that’s why I was looking at internal functions of ModelCheckpoint).

**carmocca**
The behaviour you are seeing is expected and correct. All processes run the modelcheckpoint logic and call trainer.save_checkpoint
But ultimately, only 1 checkpoint will be saved with the DDP strategy. Other strategies need to consolidate the checkpoint across ranks, hence this design

**Michał Rembalski**
But, if I were to run it with a “deepspeed” strategy, then the checkpoint that would be saved, could be partial, right? And no warning would be shown?


**Adrian**
For deepspeed it depends on the configuration, but yeah for zero3 deepspeed will save checkpoints in a folder, each rank saves one file, as this is the most important efficient way.


**Michał Rembalski**
That makes sense.
Then, If I wanted to use ddp strategy with multiple devices and store/delete checkpoint metadata:
The processes are synced before a checkpoint is created.
Should I do store/delete checkpoint metadata only on rank_zero or just “latest” one (or, if the processes are synced - it doesn’t matter, since the checkpoint is the same for all of them).
Also, how does this look for ddp_spawn? The docs say:
Since .spawn() trains the model in subprocesses, the model on the main process does not get updated.
Which process would save the checkpoint then? Would it be full/partial?


**carmocca**
 You don't need to worry about any of this! Just implement state_dict for a component and return you metadata
Note that only the state from rank zero will be saved, so if the state differs across ranks, the other rank's state will be lost. (The weights do match of course)


**Michał Tyrolski**
 - how it works behind the scene? I mean, if each gpu worker computes gradient, then only master gpu receives all gradients, accumulate it and update weights, then share it to the rest of gpus along with updates weights, is that right? In that case state_dicts will be different among different gpus, right?


**carmocca**
As you said, the gradient updates are all reduced, meaning all devices will have the same version. https://pytorch.org/tutorials/beginner/ddp_series_theory.html
Then the state dict is saved. So they will match


**Michał Tyrolski**
ok, so checkpoint is being sent only from master node, right? Btw. thx for sharing

**carmocca**
1. Each process is tied to a device in DDP
2. The devices all_reduce the weights as previously said
3.
4. Each process runs the modelcheckpoint logic
trainer.save_checkpoint() is called (maybe)
the strategy-specific logic saves the checkpoint, in the case of DDP  only rank 0 (master) does actually save
